### PR TITLE
doc: update contributing guide

### LINKED ISSUE #6: Refactor contributing guide for Notary project
The notary project has a [contributing guide](https://github.com/notaryproject/.github/blob/main/CONTRIBUTING.md) migrated from the [Notary repo](https://github.com/notaryproject/notary/blob/master/CONTRIBUTING.md). It's currently in a rough version and most of the content is outdated. 

To help new folks understand the contributing process and development workflow of the Notary project, we should consider providing a general contributing guide for the whole Notary project, including code (Notation CLI and go library repositories) and non-code (website, spec, governance) contributions.

### PR DESCRIPTION
**Pull request by yizha1**
Resolve #6 

Signed-off-by: Yi Zha <yizha1@microsoft.com>

### CODE REVIEW

File: CONTRIBUTING.md
Comment (shizhMSFT) [inline CONTRIBUTING.md:47]: How about stale issue and PRs?

File: CONTRIBUTING.md
Comment (shizhMSFT) [inline CONTRIBUTING.md:74]: nit:
```suggestion
> **Note** If you split your pull request to small changes, please make sure any of the changes goes to main will not break anything. Otherwise, it cannot be merged until this feature complete.
```

Comment (shizhMSFT) [reply CONTRIBUTING.md:74]: Renders as
> **Note** If you split your pull request to small changes, please make sure any of the changes goes to main will not break anything. Otherwise, it cannot be merged until this feature complete.

File: CONTRIBUTING.md
Comment (shizhMSFT) [inline CONTRIBUTING.md:76]: Consider `###` instead of `-`.

---

For example:

### Fork and clone
Fork the repository and clone the code to your local workspace.

### Branch
Changes should be made on your own fork in a new branch checked out from the `main` branch.

File: CONTRIBUTING.md
Comment (shizhMSFT) [inline CONTRIBUTING.md:84]: nit: You may apply [Oxford comma](https://en.wikipedia.org/wiki/Serial_comma).
```suggestion
- Develop, Build, and Test
```

File: CONTRIBUTING.md
Comment (shizhMSFT) [inline CONTRIBUTING.md:100]: > such as Closes #XXX and Fixes #XXX

Consider linking the doc at https://docs.github.com/en/issues/tracking-your-work-with-issues/linking-a-pull-request-to-an-issue#linking-a-pull-request-to-an-issue-using-a-keyword

**Review by shizhMSFT — APPROVED** [review]
LGTM

**Comment by Sumitmaithani** [comment]
I noticed that the instructions for committing and pushing changes don't include the signoff command, also known as DCO. I think it would be great to add this information to the instructions.

Currently:
```bash
git add <files_changed>
git commit -m <commit_message>
git push origin HEAD
```

Expected:
```bash
git add <files_changed>
git commit -s -m <commit_message>
git push origin HEAD
```

**Comment by Sumitmaithani** [comment]
@yizha1 👍 Looks great!

**Comment by Sumitmaithani** [comment]
Do you think it would be a good idea to update the  [CONTRIBUTING.md](https://github.com/notaryproject/notaryproject.dev/blob/main/CONTRIBUTING.md) file in this repo and add the signoff commit command? Currently, it's missing.🤔

File: CONTRIBUTING.md
Comment (FeynmanZhou) [inline CONTRIBUTING.md:91]:
```suggestion
Write code on the new branch in your fork. You can try to build the notation project locally by following the [Building guide](https://github.com/notaryproject/notation/blob/main/building.md)

Unit test cases should be added to cover the new code. For PR in [notation](https://github.com/notaryproject/notation), end-to-end (e2e) test cases should be added, see [e2e quick start](https://github.com/notaryproject/notation/blob/main/test/e2e/README.md)
```

File: CONTRIBUTING.md
Comment (FeynmanZhou) [inline CONTRIBUTING.md:99]:
```suggestion
As Notary project repositories require [DCO (Developer Certificate of Origin)](https://github.com/apps/dco) in PR workflow, contributors are required to sign off that they adhere to those requirements by adding a `Signed-off-by` line to the commit messages. Git has even provided a `-s` command line option to append that automatically to your commit messages, please use it when you commit your changes. For example:
```

File: CONTRIBUTING.md
Comment (FeynmanZhou) [inline CONTRIBUTING.md:108]: We can remind PR authors to follow the [conventional commits](https://www.conventionalcommits.org/en/v1.0.0/). This can be added in a follow-up PR.

**Review by FeynmanZhou — CHANGES_REQUESTED** [review]
PTAL

File: CONTRIBUTING.md
Comment (toddysm) [inline CONTRIBUTING.md:102]: We should have both `-s` for sign-off and `-S` for signed commits. We should also link to the instructions how to set up a private key on the machine.

Comment (FeynmanZhou) [reply CONTRIBUTING.md:102]: +1. SSH Key is much easier to set up than GPG Key from our testing. Shall we recommend contributors use SSH key to sign commits and link to https://docs.github.com/en/authentication/managing-commit-signature-verification/about-commit-signature-verification#ssh-commit-signature-verification?

File: CONTRIBUTING.md
Comment (yizha1) [reply CONTRIBUTING.md:99]: The purpose is to tell the users that notary project enforces DCO for pull requests. CI check is different.

File: CONTRIBUTING.md
Comment (yizha1) [reply CONTRIBUTING.md:102]: Thanks @toddysm. The how-to was mentioned after the command example.

`
Notary project repositories require signed commits, please refer to [commit signatures verification](https://docs.github.com/en/authentication/managing-commit-signature-verification/about-commit-signature-verification) on signing commits.`

@FeynmanZhou the above guide mentioned three options. I think users can choose the best way for them.

**Review by FeynmanZhou — APPROVED** [review]
LGTM for using it as the first version

**Review by FeynmanZhou — DISMISSED** [review]
LGTM for using it as the first version

**Comment by yizha1** [comment]
@priteshbandi @toddysm  Request re-review. We can release it as the first version if there are no big concerns, then we will iterate it over time.

File: CONTRIBUTING.md
Comment (priteshbandi) [inline CONTRIBUTING.md:31]:
```suggestion
   - Remove the 'triage' label and apply the 'question' label to indicate that this issue requires additional information. Once this issue has been clarified with the necessary details, it can then be assigned to the appropriate milestone or labeled as 'wont fix'.
```

File: CONTRIBUTING.md
Comment (priteshbandi) [inline CONTRIBUTING.md:32]:
```suggestion
   - Remove the 'triage' label and assign the 'wont fix' label to signify that this issue falls outside the scope of the Notary Project or for other relevant reasons.
```

File: CONTRIBUTING.md
Comment (priteshbandi) [inline CONTRIBUTING.md:42]:
```suggestion
A stale issue is one that remains inactive or without updates for a period of 60 days. A stale pull request (PR) is one that  remains inactive or without updates for 45 days. When an issue or PR becomes stale, it is labelled as `stale`. Normally maintainers will comment on stale issue or PR to prompt participants to take action. If there is no activity for additional 30 days, this issue or PR will be closed. If an update/comment occur on stale issues or pull requests, the stale label will be removed and the timer will restart.
```

File: CONTRIBUTING.md
Comment (shizhMSFT) [inline CONTRIBUTING.md:7]:
```suggestion
> [!Warning]
> If you are reporting a security vulnerability, Please DO NOT open a public issue.
```
it renders as

> [!Warning]
> If you are reporting a security vulnerability, Please DO NOT open a public issue.

File: CONTRIBUTING.md
Comment (shizhMSFT) [inline CONTRIBUTING.md:23]: Can an issue be converted to a discussion?

File: CONTRIBUTING.md
Comment (shizhMSFT) [inline CONTRIBUTING.md:37]: What if an issue requires multiple PRs?

File: CONTRIBUTING.md
Comment (shizhMSFT) [inline CONTRIBUTING.md:31]: What if a bug issue cannot be reproed?

File: CONTRIBUTING.md
Comment (shizhMSFT) [inline CONTRIBUTING.md:36]: Note that there are two ways to [close an issue](https://docs.github.com/en/issues/tracking-your-work-with-issues/closing-an-issue).
- Close as planned (default)
- Close as not planned

File: CONTRIBUTING.md
Comment (shizhMSFT) [inline CONTRIBUTING.md:60]: nit: `PR` --> `PRs`

File: CONTRIBUTING.md
Comment (yizha1) [reply CONTRIBUTING.md:37]: Multiple PRs need to be completed before an issue is closed

File: CONTRIBUTING.md
Comment (yizha1) [reply CONTRIBUTING.md:36]: @shizhMSFT do you mean stale issues? Any other cases that issue was closed as not planned? I don't think we need to enumerate all the scenarios, just the major scenarios are enough.

**Comment by yizha1** [comment]
@priteshbandi @shizhMSFT @FeynmanZhou would you mind reviewing this PR again? Thanks.

File: CONTRIBUTING.md
Comment (shizhMSFT) [reply CONTRIBUTING.md:37]: @yizha1 See the following picture:
image

File: CONTRIBUTING.md
Comment (shizhMSFT) [inline CONTRIBUTING.md:37]:
```suggestion
4. Issue closure: Follow the [guidance](https://docs.github.com/en/issues/tracking-your-work-with-issues/closing-an-issue)
```

**Review by FeynmanZhou — DISMISSED** [review]
I am good with this PR after the last comment get resolved. Let's get this PR merged into main ASAP and make it as the first version. 
/cc @notaryproject/notaryproject-governance-maintainers

**Review by shizhMSFT — APPROVED** [review]
LGTM

**Review by FeynmanZhou — DISMISSED** [review]
LGTM

**Review by priteshbandi — DISMISSED** [review]
LGTM

File: CONTRIBUTING.md
Comment (toddysm) [inline CONTRIBUTING.md:52]: "discuss" and "future" are a bit vague. Wouldn't labels like "in discussion" and "backlog" be better?

**Review by toddysm — CHANGES_REQUESTED** [review]

**Comment by yizha1** [comment]
Thanks for all the comments. I will resolve them before end of 12/22/2023.

File: CONTRIBUTING.md
Comment (yizha1) [reply CONTRIBUTING.md:40]: `Close as not planned` is provided by GitHub, see [guidance](https://docs.github.com/en/issues/tracking-your-work-with-issues/closing-an-issue). A screenshot for two options:

image

I have updated the texts for better explanation.

File: CONTRIBUTING.md
Comment (yizha1) [reply CONTRIBUTING.md:39]: If the author does not reply, the issue will become stale, then it will be closed as not planned eventually. If the author does not accept the decision, then this issue will not be closed, and the discussion continues in the issue. I updated the description according to GitHub guidance for simplicity.

File: CONTRIBUTING.md
Comment (yizha1) [reply CONTRIBUTING.md:45]: This practice is referred from the ORAS community. [Based on the current issue statistics](https://hackmd.io/LkKAC-5hROeEwMOPNc_D3w?view#Issues-in-last-6-months) and maintainer activities, I believe that the duration is not that long, and we can continuously improve the number over time.

For maintainers' activity, it was defined in governance guide.

**Comment by yizha1** [comment]
@priteshbandi @shizhMSFT @FeynmanZhou @toddysm Would you mind reviewing this PR again, which has been open quite a long time? 

Please be aware of some updates on the milestone naming according to Toddy's comment to bring more claritiy.

- Change `Future` milestone to `backlog` milestone
- Change `Discuss` milestone to `in dicussion` (or maybe we can remove this milestone, but label related issues and PRs `need dicussion`

**Review by shizhMSFT — APPROVED** [review]
LGTM

**Review by NiazFK — APPROVED** [review]

**Review by priteshbandi — APPROVED** [review]
LGTM

**Review by FeynmanZhou — APPROVED** [review]
LGTM

**Review by toddysm — APPROVED** [review]

**Pull request merged**
